web/app/.eslintrc

```diff
@@ -55,13 +42,47 @@
         "arrow": "parens-new-line"
       }
     ],
-    "react/no-did-update-set-state": [0],
-    "react/prefer-stateless-function": 0, // TODO: convert to stateless functions and enable
-    "react/sort-prop-types": 2,
+    "react/prop-types": [2, {
+      "ignore": ["classes"],
+      "customValidators": [],
+      "skipUndeclared": false
+    }],
     "semi": [2, "always"],
     "sort-imports": [2],
     "space-before-blocks": [2, "always"],
     "space-before-function-paren": [2, "never"],
-    "strict": [2, "safe"]
+    "strict": [2, "safe"],
+
+    // disable airbnb defaults that don't work for our project
+    // TODO: fix the code and remove these overrides
+    "arrow-body-style": 0,
```

Comment:
Ah, yes, that's a good question. The disabled rules fall into a few different categories.

There are a few rules that just won't work in our codebase. For instance:

- no-bitwise -- I guess some codebases don't need bitwise operators (and they can be confused for logical operators), but our codebase does need them for ip translation
- no-tabs -- Some of our SVGs have tabs in them
- react/jsx-props-no-spreading -- Some of our components wrap Material UI components and use spread to forward props

And then there are a few that are just stylistic preference:

- no-else-return -- This rule doesn't allow an else clause following an if clause that calls `return`; seems weird
- no-nested-ternary - We use nested ternaries all over the place
- object-curly-newline / react/jsx-curly-newline - These rules are auto-fixable but you end up with some pretty strange indentation

And there are a couple that would be great to fix, but would require some substantial rewriting of some parts of the code base (these would be great candidates for follow-up pull requests):

- no-param-reassign -- There are bunch of places where we're reassigning the value of params passed into functions
- react/no-did-update-set-state -- We're calling setState inside componentDidUpdate, and that seems bad

And finally there are a few rules that would be easy to fix but would have made this a much noisier diff (these would also be great candidates for follow-up pull requests):

- no-use-before-define -- We need to rearrange our functions so that the functions that they call are defined first.
- react/sort-comp -- We need to rearrange our components so that the lifecycle methods appear at the top.

Reply:
Thanks for the breakdown - after you merge this, I'm happy to file a (non-urgent) issue for this comment, unless you'd prefer to do it yourself.

Reply:
You're totally right, we should turn this into actionable issues. Prior to merging this branch, I'll reorganize this list into will fix / won't fix, and then I can open issues about the rules in the will fix section, as well as enabling linting for test files.